src/lib/mina_base/zkapp_command.ml

```diff
@@ -1217,81 +1236,139 @@ end = struct
+      with module Command_wrapper := Command_wrapper
+       and type cache = Cache.t = struct
+    type cache = Cache.t
+
+    let create_all (wrapped_cmds : T.t Command_wrapper.t list)
+        (init_cache : Cache.t) : t Command_wrapper.t list Or_error.t =
       Or_error.try_with (fun () ->
           snd (* remove the helper cache we folded with *)
-            (List.fold_map cmds ~init:Cache.empty
-               ~f:(fun
-                    (running_cache : Verification_key_wire.t Cache.t)
-                    { data = cmd; status }
-                  ->
+            (List.fold_map wrapped_cmds ~init:init_cache
+               ~f:(fun running_cache wrapped_cmd ->
+                 let cmd = Command_wrapper.unwrap wrapped_cmd in
+                 let cmd_failed = Command_wrapper.is_failed wrapped_cmd in
                  let verified_cmd : t =
-                   create cmd ~status ~find_vk:(fun vk_hash account_id ->
+                   create cmd ~failed:cmd_failed
+                     ~find_vk:(fun vk_hash account_id ->
                        (* first we check if there's anything in the running
                           cache within this chunk so far *)
-                       match Cache.find running_cache vk_hash with
+                       match Cache.find running_cache ~account_id ~vk_hash with
                        | None ->
-                           (* before falling back to the find_vk *)
-                           find_vk vk_hash account_id
+                           Error
+                             (Error.of_string
+                                "verification key not found in cache" )
                        | Some vk ->
                            Ok vk )
                    |> Or_error.ok_exn
                  in
                  let running_cache' =
-                   List.fold (extract_vks cmd) ~init:running_cache
-                     ~f:(fun acc vk ->
-                       Cache.set acc ~key:(With_hash.hash vk) ~data:vk )
+                   (* update the cache if the command is not failed *)
+                   if not cmd_failed then
+                     List.fold (extract_vks cmd) ~init:running_cache
+                       ~f:(fun acc (account_id, vk) ->
+                         Cache.add acc ~account_id ~vk )
+                   else running_cache
                  in
-                 (running_cache', { With_status.data = verified_cmd; status }) )
+                 ( running_cache'
+                 , Command_wrapper.map wrapped_cmd ~f:(Fn.const verified_cmd) ) )
             ) )
   end
 
-  module Any = struct
-    include Make_create_all (Map_cache)
+  (* There are 2 situations in which we are converting commands to their verifiable format:
+       - we are reasoning about the validity of commands when the sequence is not yet known
+       - we are reasoning about the validity of commands when the sequence (and by extension, status) is known
+  *)
+
+  module From_unapplied_sequence = struct
```

Comment:
Nit: The comment above reflects that this isn't a sequence, but more of a collection.